diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -543,29 +552,117 @@ bool DifferentialFunctionComparator::mayIgnore(const User *Inst) const {
+        // from the same pointer with no stores to the same pointer
+        // in between.
+        bool first = true;
+        const LoadInst *PreviousLoad = nullptr;
+        std::deque<const BasicBlock *> blockQueue = {BB};
+        std::vector<const BasicBlock *> visitedBlocks;
+        while (!blockQueue.empty()) {
+            BB = blockQueue.front();
+            blockQueue.pop_front();
+            visitedBlocks.push_back(BB);
+
+            bool searchPredecessors = true;
+            for (auto it = BB->rbegin(); it != BB->rend(); ++it) {
+                // Skip all instructions before the compared load
+                // when in the first block.
+                if (first) {
+                    if (&*it == Load) {
+                        first = false;
+                    }
+                    continue;
+                }
+
+                if (auto OrigLoad = dyn_cast<LoadInst>(&*it)) {
+                    // Try to find a previous load corresponding to the same
+                    // pointer. When found, end the seach for the current
+                    // control flow branch.
+                    if (Load->getPointerOperand()
+                        == OrigLoad->getPointerOperand()) {
+                        PreviousLoad = OrigLoad;
+                        searchPredecessors = false;
+                        break;
+                    }
+                } else if (auto Store = dyn_cast<StoreInst>(&*it)) {
+                    // Check whether a possibly conflicting store instruction
+                    // is present. If found, end the search with a failure.
+                    PreviousLoad = nullptr;
+                    searchPredecessors = false;
+                    blockQueue.clear();
+                    break;
+                }
+            }
+
+            if (searchPredecessors) {
+                auto BBPredecessors = predecessors(BB);
+                if (BBPredecessors.begin() == BBPredecessors.end()) {
+                    // If there are no more predecessors available,
+                    // end the analysis with a failure.
+                    PreviousLoad = nullptr;
+                    break;
+                }
+                // Queue up all unvisited predecessors.
+                for (auto PredBB : BBPredecessors) {
+                    if (std::find(blockQueue.begin(), blockQueue.end(), PredBB)
+                                == blockQueue.end()
+                        && std::find(visitedBlocks.begin(),
+                                     visitedBlocks.end(),
+                                     PredBB)
+                                   == visitedBlocks.end()) {
+                        blockQueue.push_back(PredBB);
                     }
                 }
-                return true;
             }
         }
-        return false;
+        // If the load is repeated without stores in between, skip
+        // it because it creates no semantic differences.
```

Comment:
better comment: `... because it is redundant and its removal will cause no semantic difference.`